h5py/h5t.pyx

```diff
@@ -1421,27 +1421,61 @@ cdef TypeCompoundID _c_complex(dtype dt):
 
     return TypeCompoundID(tid)
 
-cdef TypeCompoundID _c_compound(dtype dt, int logical):
+cdef TypeCompoundID _c_compound(dtype dt, int logical, int aligned):
     # Compound datatypes
 
     cdef hid_t tid
     cdef TypeID type_tmp
     cdef dtype dt_tmp
     cdef size_t offset
+    cdef size_t offset_step = 0
 
-    cdef dict fields = dt.fields
     cdef tuple names = dt.names
-
-    # Initial size MUST be 1 to avoid segfaults (issue 166)
-    tid = H5Tcreate(H5T_COMPOUND, 1)  
-
-    for name in names:
-        ename = name.encode('utf8') if isinstance(name, unicode) else name
-        dt_tmp = dt.fields[name][0]
-        offset = dt.fields[name][1]
-        type_tmp = py_create(dt_tmp, logical=logical)
-        H5Tset_size(tid, offset+type_tmp.get_size())
-        H5Tinsert(tid, ename, offset, type_tmp.id)
+    cdef dict fields = {}
+    cdef list offsets
+
+    # The challenge with correctly converting a numpy/h5py dtype to a HDF5 type
+    # which is composed of subtypes has three aspects we must consider
+    # 1. numpy/h5py dtypes do not always have the same size and HDF5, even when
+    #   equivalent (can result in overlapping elements if not careful)
+    # 2. For correct round-tripping of aligned dtypes, we need to consider how
+    #   much padding we need
+    # 3. There is no requirement that the offsets be monotonically increasing
+    #
+    # The code below tries to cover these aspects
+
+    for name, field in dt.fields.items():
+        dt_tmp = field[0]
+        offset = field[1]
+        fields[offset] = {
+            "name": name.encode('utf8') if isinstance(name, unicode) else name,
```

Comment:
this works because it's cython but won't work on Python (excluding Python 2) where `unicode` is not defined.

Reply:
Hey there, I've been recently working on "structured datatype" improvements in the numpy project, with changes incoming in numpy 1.14. Previously both the behavior and documentation of structured datatypes had problems and ommisions. See the discussion at the end of https://github.com/numpy/numpy/pull/6053. I've also heavily updated the structured datatype docs, [rendered here]https://ahaldane.github.io/user/basics.rec.html] for now.

This loop should loop over `dt.names`, rather than over `dt.fields.items`, for two reasons: First, `dt.names` is a list, so preserves field order, whille `dt.fields` is dict like, so does not preserve order, and second, `dt.fields` also may have "field titles", which you can read about [here](https://ahaldane.github.io/user/basics.rec.html#field-titles), which will mess up the loop. See in particular the admonition in that section about always looping over `names` instead of `fields`. Especially after numpy 1.14, field order matters (and it also matters in the HDF5 spec).